File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ObsoleteAttributesNeedExplanation.vb
' Commented line for concurrent namespace

<Obsolete>  ' Noncompliant^2#8 {{Add an explanation.}}
Class Noncompliant
    <Obsolete()> Sub WithBrackets() ' Noncompliant {{Add an explanation.}}
'    ^^^^^^^^^^
    End Sub

    <System.Obsolete> ' Noncompliant
    Sub FullyDeclaredNamespace()
    End Sub

    <Global.System.Obsolete> ' Noncompliant
    Sub GloballyDeclaredNamespace()
    End Sub

    <Obsolete(Nothing)> ' Noncompliant
    Sub WithNothing()
    End Sub

    <Obsolete("")> ' Noncompliant
    Sub WithEmptyString()
    End Sub

    <Obsolete("  ")> ' Noncompliant
    Sub WithWhiteSpace()
    End Sub

    <Obsolete("", True)> ' Noncompliant
    Sub WithTwoArguments()
    End Sub

    <Obsolete> ' Noncompliant
    <CLSCompliant(False)>
    Function Multiple() As UInteger
        Return 0
    End Function

    <Obsolete, CLSCompliant(False)> Function Combined() As UInteger ' Noncompliant
        Return 0
    End Function

    <Obsolete> ' Noncompliant
    Enum [Enum]
        foo
        bar
    End Enum

    <Obsolete> ' Noncompliant
    Sub New()
    End Sub

    <Obsolete> ' Noncompliant
    Sub Method()
    End Sub

    <Obsolete> ' Noncompliant
    Property [Property] As Integer

    <Obsolete> ' Noncompliant
    Private Field As Integer

    <Obsolete> ' Noncompliant
    Event [Event] As EventHandler

    <Obsolete> ' Noncompliant
    Delegate Sub [Delegate]()
End Class

<Obsolete> ' Noncompliant
Interface IInterface
    <Obsolete> ' Noncompliant
    Sub Method()
End Interface

<Obsolete> ' Noncompliant
Structure ProgramStruct
    <Obsolete> ' Noncompliant
    Sub Method()
    End Sub
End Structure

<Obsolete("explanation")>
Class Compliant
    <Obsolete("explanation")>
    Enum [Enum]
        foo
        bar
    End Enum

    <Obsolete("explanation")>
    Sub New()
    End Sub

    <Obsolete("explanation")>
    Sub Method()
    End Sub

    <Obsolete("explanation", True)>
    Sub WithTwoArguments()
    End Sub

    <Obsolete("explanation")>
    Property [Property] As String

    <Obsolete("explanation", True)>
    Private Field As Integer

    <Obsolete("explanation", False)>
    Event [Event] As EventHandler

    <Obsolete("explanation")>
    Delegate Sub [Delegate]()
End Class

<Obsolete("explanation")>
Interface IComplaintInterface
    <Obsolete("explanation")>
    Sub Method()
End Interface

<Obsolete("explanation")>
Structure ComplaintStruct
    <Obsolete("explanation")>
    Sub Method()
    End Sub
End Structure

Class NotApplicable
    <CLSCompliant(False)>
    Enum [Enum]
        foo
        bar
    End Enum

    Sub New()
    End Sub

    Sub Method()
    End Sub

    Property [Property] As Integer

    Private Field As Integer

    Event [Event] As EventHandler

    Delegate Sub [Delegate]()

    <NotSystem.ObsoleteAttribute>
    Sub SameName()
    End Sub
End Class

Namespace NotSystem
    Class ObsoleteAttribute
        Inherits Attribute
    End Class
End Namespace

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ObsoleteAttributesNeedExplanation.cs
using System;

[Obsolete] // Noncompliant ^2#8 {{Add an explanation.}}
class Noncompliant
{
    [Obsolete()] // Noncompliant
    void WithBrackets() { }

    [System.Obsolete] // Noncompliant
    void FullyDeclaredNamespace() { }

    [global::System.Obsolete] // Noncompliant
    void GloballyDeclaredNamespace() { }

    [Obsolete(null)] // Noncompliant
    void WithNull() { }

    [Obsolete("")] // Noncompliant
    void WithEmptyString() { }

    [Obsolete("  ")] // Noncompliant
    void WithWhiteSpace() { }

    [Obsolete("", true)] // Noncompliant
    void WithTwoArguments() { }

    [Obsolete] // Noncompliant
    [CLSCompliant(false)]
    uint Multiple() { return 0; }

    [Obsolete, CLSCompliant(false)]
//   ^^^^^^^^
    uint Combined() { return 0; }

    [Obsolete] // Noncompliant
    enum Enum { foo, bar }

    [Obsolete] // Noncompliant
    Noncompliant() { }

    [Obsolete] // Noncompliant
    void Method() { }

    [Obsolete] // Noncompliant
    int Property { get; set; }

    [Obsolete] // Noncompliant
    int Field;

    [Obsolete] // Noncompliant
    event EventHandler Event;

    [Obsolete] // Noncompliant
    delegate void Delegate();
}

[Obsolete] // Noncompliant
interface IInterface
{
    [Obsolete] // Noncompliant
    void Method();
}


[Obsolete] // Noncompliant
struct ProgramStruct
{
    [Obsolete] // Noncompliant
    void Method() { }
}

[Obsolete("explanation")]
class Compliant
{
    [Obsolete("explanation")]
    enum Enum { foo, bar }

    [Obsolete("explanation")]
    Compliant() { }

    [Obsolete("explanation")]
    void Method() { }

    [Obsolete("explanation", true)]
    void WithTwoArguments() { }

    [Obsolete("explanation")]
    string Property { get; set; }

    [Obsolete("explanation", true)]
    int Field;

    [Obsolete("explanation", false)]
    event EventHandler Event;

    [Obsolete("explanation")]
    delegate void Delegate();
}

[Obsolete("explanation")]
interface IComplaintInterface
{
    [Obsolete("explanation")]
    void Method();
}

[Obsolete("explanation")]
struct ComplaintStruct
{
    [Obsolete("explanation")]
    void Method() { }
}

class Ignore
{
    // FP the value of error is taken.
    [Obsolete(error: true, message: "explanation")] // Noncompliant
    public void NamedParmetersDifferentOrderFP() { }

    [Obsolete(error: true, message: "")] // Noncompliant for wrong reason 
    public void NamedParmetersDifferentOrder() { }
}

class NotApplicable
{
    [CLSCompliant(false)]
    enum Enum { foo, bar }

    NotApplicable() { }

    void Method() { }

    int Property { get; set; }

    int Field;

    event EventHandler Event;

    delegate void Delegate();

    [NotSystem.Obsolete]
    void SameName() { }
}

namespace NotSystem
{
    public class ObsoleteAttribute : Attribute { }
}
